tutorials/week4-react.md

```diff
@@ -0,0 +1,511 @@
+
+A few things to note from the above example:
+
+- Component.defaultProps can be used to specify default values for props.
+- Components are rendered to the user interface and the component’s logic contains the data to be displayed in the view(UI).
+- Curly braces ({}) in JSX/TSX are used as a connection between the data to be displayed in the view and the component’s logic for displaying the data in the view.
+  - In our example, `{props.name}` will reflect the value of the property `name` in the view(html) for the cases when the values "John" and "Jane" are passed as props for the `name` property.
+  - If no props are passed for an instance of the component, then it will display the default value of props.
+
+## State 
+State management is just a means of facilitating data sharing and communication among components. It creates a concrete data structure that you can read and write to reflect the state of your program.
+
+```ts
+const [counter, setCounter] = useState(0)
+```
+
+The above snippet shows creation of counter state with an intial value of 0. Using the [array destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) syntax we extract out the state variable and the function to update the counter value. 
+
+In its most basic form, a State object is a JavaScript object that represents the part of a component that can change as a result of a user's action. States can also be thought of as a component's memory.
+
+State update calls are asynchronous. As one cannot expect to call the update state function on one line and expect the state to be updated on the next. The reason for this is because update state methods are more of a request than an immediate order to update state. So React schedules an update to a component’s state object. When state changes, the component responds by re-rendering. Also multiple update request may be batched into one for performance reasons. 
+
+Changes in state and/or props will both cause our React component to re-render. Changes in state, on the other hand, can only occur internally as a result of components modifying their own state. 
+
+## Communication Between Components: 
+
+### Parent To Child Component Communication: 
+<a href="https://codesandbox.io/s/nervous-morse-o3pwqm?file=/src/tutorial/ParentChildCommunication.tsx" target="_blank">view in sandbox</a>
+
+
+Passing values from a parent component to a child component is simple. We only have to pass the values as props of the child component. The child component can then use the props object to output results. In the example code you will see that CounterContent component accepts a counter prop which is then used to display the value inside div element. 
+```ts
+import { useState } from "react";
+
+interface CounterContentProps {
+  counter: Number;
+}
+
+function CounterContent({ counter }: CounterContentProps) {
+  return <div>Counter: {counter}</div>;
+}
+
```

Comment:
Throws error: 

<img width="479" alt="image" src="https://github.com/neu-se/CS4530-Fall-2023/assets/42910418/6a28fdc7-959b-4ee0-8b52-485a0fcb7bdf">


we can typecast to string: 

`  return <div>Counter: { counter.toString() }</div>;`